pkg/util/expansion/expand.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package expansion
+
+const (
+	operator        = '$'
+	referenceOpener = '('
+	referenceCloser = ')'
+)
+
+// syntaxWrap returns the input string wrapped the expansion syntax.
+func syntaxWrap(input string) string {
+	return string(operator) + string(referenceOpener) + input + string(referenceCloser)
+}
+
+// MappingFuncFor returns a mapping function for use with Expand that
+// implements the expansion semantics defined in the expansion spec; it
+// returns the input string wrapped in the expansion syntax if no mapping
+// for the input is found.
+func MappingFuncFor(context ...map[string]string) func(string) string {
+	return func(input string) string {
+		for _, vars := range context {
+			val, ok := vars[input]
+			if ok {
+				return val
+			}
+		}
+
+		return syntaxWrap(input)
+	}
+}
+
+// Expand replaces variable references in the input string according to
+// the expansion spec using the given mapping function to resolve the
+// values of variables.
+func Expand(input string, mapping func(string) string) string {
+	buf := make([]byte, 0, 2*len(input))
```

Comment:
@bgrant0607 10-4
On Thu, May 14, 2015 at 6:24 PM Brian Grant <notifications@github.com>
wrote:

> In pkg/util/expansion/expand.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7936#discussion_r30369819>
> :
>
> > +		for _, vars := range context {
> > +			val, ok := vars[input]
> > +			if ok {
> > +				return val
> > +			}
> > +		}
> > +
> > +		return syntaxWrap(input)
> > +	}
> > +}
> > +
> > +// Expand replaces variable references in the input string according to
> > +// the expansion spec using the given mapping function to resolve the
> > +// values of variables.
> > +func Expand(input string, mapping func(string) string) string {
> > +	buf := make([]byte, 0, 2*len(input))
>
> Please put it under third_party/golang. Since it's no longer env var
> expansion, it doesn't need to go under "os". "expansion" would be fine. It
> should mention which file/package it was cribbed from, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7936/files#r30369819>
> .
>